File: epi_cpp/01_PrimitiveTypes/14_uniform_random.cpp
#include <iostream>
#include <random>

using namespace std;

int UniformRandom(int lower_bound, int upper_bound){
    int number_of_outcomes = upper_bound - lower_bound + 1;
    int result;
    default_random_engine seed((random_device())());
    uniform_int_distribution<int> distro(0, 1);
    do{
        result = 0;
        for(int i=0; (1<<i) < number_of_outcomes; ++i){
            result = (result<<1)|distro(seed);
        }
    }while(result >= number_of_outcomes);
    return result;
}

int main(){
    int lower_bound = 3, upper_bound = 20;
    for(int i=0; i < 10; ++i){
        cout << UniformRandom(lower_bound, upper_bound) << endl;
    }
    return 0;
}

File: epi_python/01_PrimitiveTypes/14_uniform_random.py
import random

def uniform_random(lower_bound: int, upper_bound: int)->int:
    number_of_outcomes = upper_bound - lower_bound + 1
    while True:
        result, i = 0, 0
        while (1<<i) < number_of_outcomes:
            result = (result<<1)|random.randint(0,1)
            i += 1
        if result < number_of_outcomes:
            break
    return result + lower_bound

if __name__ == '__main__':
    print(uniform_random(10, 20))